src/ui/public/fancy_forms/kbn_form_controller.js

```diff
@@ -1,74 +1,81 @@
-import _ from 'lodash';
+export function decorateFormController($delegate, $injector) {
+  const [directive] = $delegate;
+  const FormController = directive.controller;
 
-/**
- * Extension of Angular's FormController class
- * that provides helpers for error handling/validation.
- *
- * @param {$scope} $scope
- */
-export function KbnFormController($scope, $element) {
-  const self = this;
+  class KbnFormController extends FormController {
+    // prevent inheriting FormController's static $inject property
+    // which is angular's cache of the DI arguments for a function
+    static $inject = ['$scope', '$element'];
 
-  self.errorCount = function () {
-    return self.$$invalidModels().length;
-  };
+    constructor($scope, $element, ...superArgs) {
+      super(...superArgs);
 
-  // same as error count, but filters out untouched and pristine models
-  self.softErrorCount = function () {
-    return self.$$invalidModels(function (model) {
-      return model.$touched || model.$dirty;
-    }).length;
-  };
-
-  self.describeErrors = function () {
-    const count = self.softErrorCount();
-    return count + ' Error' + (count === 1 ? '' : 's');
-  };
-
-  self.$$invalidModels = function (predicate) {
-    predicate = _.callback(predicate);
+      $element.on('submit', this._onSubmit);
+      $scope.$on('$destroy', () => {
+        $element.off('submit', this._onSubmit);
+      });
+    }
 
-    const invalid = [];
+    errorCount() {
+      return this._getInvalidModels().length;
+    }
 
-    _.forOwn(self.$error, function collect(models) {
-      if (!models) return;
+    // same as error count, but filters out untouched and pristine models
+    softErrorCount() {
+      return this._getInvalidModels()
+        .filter(model => model.$touched || model.$dirty)
+        .length;
+    }
 
-      models.forEach(function (model) {
-        if (model.$$invalidModels) {
-          // recurse into child form
-          _.forOwn(model.$error, collect);
-        } else {
-          if (predicate(model)) {
-            // prevent dups
-            let len = invalid.length;
-            while (len--) if (invalid[len] === model) return;
+    describeErrors() {
+      const count = this.softErrorCount();
+      return `${count} Error${count === 1 ? '' : 's'}`;
+    }
 
-            invalid.push(model);
-          }
-        }
-      });
-    });
+    $setTouched() {
+      this._getInvalidModels()
+        .forEach(model => model.$setTouched());
+    }
 
-    return invalid;
-  };
+    _onSubmit = () => {
+      if (this.errorCount()) {
+        event.preventDefault();
+        event.stopImmediatePropagation();
+        this.$setTouched();
+      }
+    }
 
-  self.$setTouched = function () {
-    self.$$invalidModels().forEach(function (model) {
-      // only kbnModels have $setTouched
-      if (model.$setTouched) model.$setTouched();
-    });
-  };
+    _getInvalidModels() {
```

Comment:
Can you break the definition of this function down into a couple variables with useful names?  There's a lot of logic nested in the next 10 or so lines, and I think some variables may help.

I also recommend moving the declaration of collect() just outside the class or something rather than constructing it inline as a function expression like this.  As far as I can tell, collect() doesn't rely on any class state, so I think it can be moved pretty seamlessly.